verl/model_merger/megatron_model_merger.py

```diff
@@ -0,0 +1,369 @@
+from safetensors.torch import load_file
+from transformers import (
+    AutoConfig,
+    PretrainedConfig,
+)
+
+from verl.models.mcore import hf_to_mcore_config
+from verl.utils.device import get_nccl_backend
+from verl.utils.megatron.dist_checkpointing import load_dist_checkpointing
+from verl.utils.megatron_utils import get_model
+
+from .base_model_merger import BaseModelMerger, ModelMergerConfig
+
+
+@contextmanager
+def noop_context() -> Any:
+    yield
+
+
+class MegatronModelMerger(BaseModelMerger):
+    """
+    Model merger for Megatron-LM distributed checkpoints.
+
+    This class handles the conversion of Megatron-LM distributed checkpoints into HuggingFace format.
+    Megatron-LM uses tensor parallelism, pipeline parallelism, and data parallelism to distribute
+    large language models across multiple GPUs. This merger reconstructs the full model by
+    loading distributed checkpoints and applying the necessary transformations.
+
+    Key features:
+    - Support for tensor parallel, pipeline parallel, and data parallel configurations
+    - Automatic parameter name mapping from Megatron to HuggingFace conventions
+    - Handling of QKV and gate-up tensor splitting/merging
+    - Support for tied word embeddings and value models
+    - Integration with Megatron's distributed checkpointing system
+
+    The merger handles various model architectures and configurations:
+    - Standard transformer models (GPT-style)
+    - Models with tied word embeddings
+    - Value models for reinforcement learning
+    - Multi-layer attention (MLA) architectures
+    - Mixture of Experts (MoE) models
+
+    Args:
+        config (ModelMergerConfig): Configuration object with Megatron-specific settings
+            including tie_word_embedding and is_value_model flags.
+
+    Example:
+        To merge Megatron checkpoints:
+        ```python
+        config = ModelMergerConfig(
+            operation="merge",
+            backend="megatron",
+            local_dir="path/to/megatron/checkpoints",
+            target_dir="path/to/output",
+            tie_word_embedding=True
+        )
+        merger = MegatronModelMerger(config)
+        merger.merge_and_save()
+        ```
+    """
+
+    def __init__(self, config: ModelMergerConfig):
+        from verl.utils.megatron_utils import get_hf_config_and_tokenizer_checkpoint_path
+
+        config.hf_model_config_path = get_hf_config_and_tokenizer_checkpoint_path(config.local_dir)
+        super().__init__(config)
+        os.environ["RANK"] = "0"
+        os.environ["WORLD_SIZE"] = "1"
+        os.environ["MASTER_ADDR"] = "localhost"
+        os.environ["MASTER_PORT"] = "12355"
+        torch.distributed.init_process_group(get_nccl_backend())
+        mpu.initialize_model_parallel(
+            tensor_model_parallel_size=1,
+            virtual_pipeline_model_parallel_size=None,
+            context_parallel_size=1,
+            expert_model_parallel_size=1,
+        )
```

Comment:
add one comment for single rank distributed set up and multiple rank loading is currently not supported. in the future consider warp environment into a reusable method.